pyrobosim/pyrobosim/navigation/astar_grid.py

```diff
@@ -0,0 +1,264 @@
+        self.visited = set()  # (x, y)
+        self.start = None
+        self.goal = None
+        self.num_nodes_expanded = 0
+        self.distance_metric = distance_metric
+        self.diagonal_motion = diagonal_motion
+        self.actions = {
+            "left": {"cost": 1, "action": (-1, 0)},
+            "right": {"cost": 1, "action": (1, 0)},
+            "up": {"cost": 1, "action": (0, 1)},
+            "down": {"cost": 1, "action": (0, -1)},
+            "left_up": {"cost": 1.414, "action": (-1, 1)},
+            "left_down": {"cost": 1.414, "action": (-1, -1)},
+            "right_up": {"cost": 1.414, "action": (1, 1)},
+            "right_down": {"cost": 1.414, "action": (1, -1)},
+        }
+        keys = list(self.actions.keys())
+        self.selected_actions = keys if self.diagonal_motion else keys[:4]
+        print(f"Selected actions : {self.selected_actions}")
+
+        # Set the heuristic function
+        if self.distance_metric == "Manhattan":
+            self._heuristic = lambda p1, p2: abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
+        elif self.distance_metric == "Euclidian":
+            # Not taking the square root since we dont need the actual distance
+            self._heuristic = lambda p1, p2: (p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2
+        print(f"Selected heuristic: {self.distance_metric}")
+
+    def _reset(self):
+        """Resets the state of data used by the algorithm"""
+        self.grid = occupancy_grid_from_world(
+            self.world,
+            resolution=self.resolution,
+            inflation_radius=self.inflation_radius,
+        )
+        self.candidates = PriorityQueue()
+        self.visited.clear()
+        self.goal = None
+        self.start = None
+        self.latest_path = Path()
+        self.num_nodes_expanded = 0
+
+    def _expand(self, node):
+        """
+        Generates and adds free neighbours to exploration candidates
+        :param node: The node to be expanded from
+        :type node: :class: Node
+        """
+        for action_name in self.selected_actions:
+            cost = self.actions[action_name]["cost"]
+            action = self.actions[action_name]["action"]
+            _x = node.x + action[0]
+            _y = node.y + action[1]
+            if (_x, _y) not in self.visited and not self.grid.is_occupied(_x, _y):
+                _g = node.g + cost
+                _h = self._heuristic((_x, _y), self.goal)
+                self.candidates.put(Node(_x, _y, _g, _h, parent=node), block=False)
+                self.num_nodes_expanded += 1
+
+    def _get_best_candidate(self):
+        """Return the candidate with best metric"""
+        return self.candidates.get()
+
+    def _mark_visited(self, node):
+        """
+        Adds the given node to visited nodes collection
+        :param node: The node that has been visited
+        :type node: :class: Node
+        """
+        self.visited.add((node.x, node.y))
+
+    def _generate_path(self, node):
```

Comment:
I had another fun idea here.

Would be cool if generating the path could club together steps into longer straight lines; for example, if you have a path segment that moves right for 25 grid cells, instead of adding 25 individual points, it could all be lumped as one segment.

Reply:
That will be good to have !